### ISSUE TITLE
host_callback primitives are not usable under xla.lower_fun

### ISSUE THREAD

**Issue by gnecula** [jax-ml/jax#8853]
This concerns an unusual use of host_callback, inside a custom primitive whose XLA lowering rule uses `xla.lower_fun`.
This uses private JAX core APIs and your mileage may vary.

This usage is not working properly when the function passed to `lower_fun` includes host callback primitives.

```
    class Prim(jax.core.Primitive):
      def __init__(self):
        super(Prim, self).__init__("prim")
        self.multiple_results = True

        def _abstract(*flat_avals, **params):
          return pe.abstract_eval_fun(self.impl, *flat_avals, **params)

        self.def_abstract_eval(_abstract)

        def _xla(c, *xla_args, **params):
          translation = xla.lower_fun(self.impl, multiple_results=True,
                                      backend="cpu")
          return translation(c, *xla_args, **params)

        xla.translations[self] = _xla

      def impl(self, *args):
        x, = args
        ret = hcb.id_print(x * 3., output_stream=testing_stream)
        return ret,

    @jax.jit
    def fn(x):
      return Prim().bind(x)[0]

    res = fn(jnp.asarray(1.))
```

This crashes with `TypeError: _outside_call_translation_rule() missing 1 required keyword-only argument: 'has_token'`

**Comment by gnecula** [comment]
In #8841 I have improved the situation, but there is still a problem. If the `lower_fun` calls to a host_callback primitive and there are other host_callback primitives in the surrounding code then the sequencing tokens are not passed properly through the inner primitives. This may have the effect that the inner host_callback will execute out of order w.r.t. the outer primitives. 

Fixing this requires a more thorough handing of effects in JAX.

**Comment by lukemetz** [comment]
Does this out of order issue mean incorrect results in this case? Would you advise against patching this PR into a local copy and running with that?

I really would like to be able to have a custom grad + custom batching rule for a host_callback call and all my workaround attempts keep failing. Any input appreciated. Thanks!

**Comment by gnecula** [comment]
The problem arises when you have a computation that uses hcb primitives directly and also under the lower_fun. In that case the operations needed for the implementation of hcb primitives will not be sequenced properly and they may execute out of order. Most likely this will result in crashes. I don't think that this can result in incorrect results, but I am not sure.


I did not realize that you are doing this to get a custom batching rule for hcb.call. (A custom grad will work if you use the public APIs for custom grads.)

**Comment by lukemetz** [comment]
Good to know! Thanks!

For what it's worth I believe I found a workaround to our original problem -- instead of a custom batching rule just define a batching rule for all outside_call_p and have this rule hardcoded for the one particular hcb.call we use. This won't work with more than one hcb / arbitrary numbers of hcb.call but that is fine with me. 

Thanks again for your help on this issue.

**Comment by patrick-kidger** [comment]
Any movement on this issue? I've just bumped up against it when using custom primitives to define a custom batching rule for a `host_callback.call`.

(FWIW `custom_batching.custom_vmap` doesn't seem to solve my problem either -- Idk to what extent that's intended to be usable at the moment but with that approach I get `NotImplementedError: outfeed rewrite custom_vmap_call` when trying to JIT.)

**Comment by patrick-kidger** [comment]
Fair enough! For completeness I managed to work around this (to produce a `host_callback.call` that can be both vmap'd and jit'd) by adapting @lukemetz's solution and [monkey-patching the batching rule](https://github.com/patrick-kidger/equinox/blob/3af0cdefd70e7a36850260a34cbc58d52c6a4b59/equinox/experimental/stateful.py#L46) for `outside_call_p`. (With a bit of care to only activate in my use case, and not overwrite the existing batching rule for any other case.)

**Comment by froystig** [comment]
Current state of `custom_vmap` is, roughly:

* some but not all transformations are supported (mainly we have JVP but no reverse-mode AD);
* it is currently "initial style", which we may want to change;
* it doesn't yet support self-reference (e.g. the custom batching rule can't call the original custom-vmapped function)

I was holding on documenting and declaring it ready for wide use until these are complete/decided. Do we know that having it will resolve this issue, or do we have enough in place to try and find out?
